src/odemis/util/test/img_test.py

```diff
@@ -1181,8 +1183,825 @@ def test_rotate_img_metadata(self):
+        # Verify that using a CTZ=111 5D DA will return the right 2D DA.
+        out_arr = img.ensure2DImage(self.data)
+
+        # check if the YX values match
+        self.assertEqual(out_arr.shape, self.data.shape[3:])
+
+        # Verify that using a full 5D DA will fail and see if a ValueError exception is raised
+        data = self.data.reshape(1, 2, 50, 400, 400)
+
+        with self.assertRaises(ValueError):
+            img.ensure2DImage(data)
+
+
+class TestSubtract(unittest.TestCase):
+
+    def setUp(self):
+        self.pic1 = numpy.zeros((600, 600))
+        self.pic2 = numpy.zeros((600, 600))
+
+    def test_subtract_uint(self):
+        """
+        Verify that using an uint8 as dtype will return the right DA.
+        """
+        pic1 = self.pic1.astype('uint8')
+        pic2 = self.pic2.astype('uint8')
+
+        # add white squares in the black pictures
+        pic1[200:500, 200:500] = 255
+        pic2[300:400, 300:400] = 255
+
+        # subtract the pictures
+        out_arr = img.Subtract(pic1, pic2)
+
+        # output shape should be the same as biggest input DA
+        self.assertEqual(out_arr.shape, pic1.shape)
+        # all values in this square should return false=0
+        self.assertTrue(numpy.all(out_arr[300:400, 300:400] == 0))
+
+    def test_subtract_underflow(self):
+        """
+        Verify that using an uint8 as dtype will return the right DA.
+        Also test that underflow does not happen.
+        """
+        pic1 = self.pic1.astype('uint8')
+        pic2 = self.pic2.astype('uint8')
+
+        # 10 -11 == -1, but uint8 doesn't support it, so it should be clipped to 0.
+        pic1[:] = 10
+        pic2[:] = 11
+
+        # subtract the pictures
+        out_arr = img.Subtract(pic1, pic2)
+
+        # output shape should be the same as biggest input DA
+        self.assertEqual(out_arr.shape, pic1.shape)
+        # see if the return value is of type uint8
+        self.assertEqual(out_arr.dtype, numpy.dtype('uint8'))
+        # all values in this square should return false=0
+        numpy.testing.assert_equal(out_arr, 0)
+
+    def test_subtract_float(self):
+        """
+        Verify that using a float as dtype will return the right DA.
+        """
+        pic1 = self.pic1.astype('float32')
+        pic2 = self.pic2.astype('float32')
+
+        # add white squares in the black pictures
+        pic1[200:500, 200:500] = 0.5
+        pic2[300:400, 300:400] = 0.5
+
+        # subtract the pictures
+        out_arr = img.Subtract(pic1, pic2)
+
+        # output shape should be the same as biggest input DA
+        self.assertEqual(out_arr.shape, pic1.shape)
+        # all values in this square should return false=0
+        self.assertTrue(numpy.all(out_arr[300:400, 300:400] == 0))
+
+    def test_subtract_uint_float(self):
+        """
+        Verify that using pictures with different dtypes will return the right DA.
+        Check if using 2 different dtype arrays will still work.
```

Comment:
Actually, `Subctract()` docinfo says  "return : a - b, with same dtype [...] as a".
So uint8 - float32 should yield an uint8, but it returns a float.

=> For now write down that this doesn't work. I'd suggest to also just have explicit check on the dtype.
```python
        # FIXME: for now this doesn't pass because Subtract() always return a float
        out_arr = img.Subtract(pic1, pic2)
        # self.assertEqual(out_arr.dtype, pic1.dtype)

        out_arr = img.Subtract(pic2, pic1)
        self.assertEqual(out_arr.dtype, pic2.dtype)
```